### ISSUE TITLE
Add ability to make a safe C call with its own stack frame

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#799]
`duk_safe_call()` performs a protected C call, but doesn't give the function its own stack frame.  The safe call can thus access and pop values beyond those it was allocated.  Sometimes that's useful, however sometimes you just don't trust the function you're calling at all because it's at arm's length from your main logic (callbacks, for example), and it would nice to give them their own stack frame like Duktape/C calls.

I propose an API (or perhaps a flag for `duk_safe_call()`) which would allow this.

This could even be used internally, for example to protect the debugger's app request callback, as mentioned in #796.

**Comment by svaarala** [comment]
There are three main issues related to making safe calls:

- Is the call part of the call stack?
- Is there a function object in the Ecmascript space to represent the target? This is tied to how the call appears in the call stack.
- Is the value stack frame shifted or not?

Current safe call doesn't appear in the call stack, doesn't have a representative function objcet, and doesn't shift the value stack.

A minimal change might be to keep things otherwise as is, but shift the value stack frame only. This may be trivial to implement (or not); call stack information is used to restore a value stack shape on function returns but since `duk_safe_call()` is protected, it can't be skipped in unwind and should thus be able to unshift the value stack correctly.

I agree this might be cleaner behavior than the current. And one could still access the entire value stack frame simply by giving suitable arguments: give nargs as `duk_get_top()` to make the value stack shift a no-op for example.

**Comment by fatcerberus** [comment]
I think I like the minimal solution (shift stack frame only) most.  Giving it a full Ecmascript callstack entry is overkill because the call is logically still confined to C, you're just piggybacking on Duktape's error handling mechanism to protect the call.

**Comment by svaarala** [comment]
I agree, though I'd prefer to just deprecate the previous safe call model to avoid maintaining and documenting two different behaviors. API documentation and migration documentation should then explain how to emulate the previous behavior manually (e.g. use `duk_get_top()` as nargs and some other suitable value as `nrets`).

**Comment by svaarala** [comment]
So I guess the new behavior would be:

- `nargs` identifies how many value stack elements are captured in the new value stack frame. If the current frame is smaller, API error.
- `nrets` identifies how many value stack elements remain on the value stack after the call; in other words, the final value stack size would be `duk_get_top() - nargs + nrets`. The `duk_get_top() - nargs` index would be the bottom of the new value stack frame.

Safe call functions which operated only with negative indices would work without change. Safe call functions which accessed to the whole value stack or used positive indices would need to be fixed, or the `duk_safe_call()` call site tweaked to emulate the previous behavior.

**Comment by svaarala** [comment]
It might be useful to rename the API call at the same time because its semantics would be a bit different than the previous call (adding the userdata wasn't an issue in this way). Maybe `duk_protected_call()`?

This would make it clear that `duk_safe_call()` sites need to be converted, and would also allow `extras/duk-v1-compat` to include a `duk_safe_call()` which would be (at least mostly) API compatible, including not having the userdata argument.

**Comment by fatcerberus** [comment]
I imagine the majority of safe calls in use will be using negative indices (the documentation specifically recommends doing so), so it shouldn't be too much of a migration burden in practice.

Agree on the rename, and "protected" matches other calls which use that terminology, e.g. `duk_pcall` ("protected call") or `duk_pcompile` ("protected compile").  Consistency is good. :)

**Comment by fatcerberus** [comment]
One thing that would be nice about this is for calls like the aforementioned app request callback: the callback wouldn't have to get a separate `nargs` and could discover the number of values by simply calling `duk_get_top()`.

In this way the callback could be made to have the same signature as a standard protected call and called directly using the new API.  The value stack is protected and the callback can use positive value stack indices--everybody wins. :)

**Comment by fatcerberus** [comment]
Related to the above: Could there also be a way to support variable number of return values?

**Comment by svaarala** [comment]
That should be possible, and will also be needed for the ES6 multi-return feature eventually. In both cases it's important that the callee indicate how many return values exist and the caller to indicate how many it wants to see. (Also for ES6, it'd be nice to be able to indicate if the caller wants the return values in an array or expanded on the value stack.)

A small practical issue is that if the only return value is a success/failure one, the variable return value count needs to be computed by the caller as `duk_get_top() - pre_call_top - nargs`.

**Comment by svaarala** [comment]
`DUK_VARARGS` for argument count and `DUK_VARRET` for return count? Or maybe `DUK_VARARGS` for both? Maybe rename `DUK_VARARGS` to something more neutral (?)

It would be nice to sync this change with the ES6 multireturn support. I've been considering adding multireturn support to the C API (before Ecmascript destructuring assignment) so that you'd be able to say `return 3;` in a Duktape/C function, and when making calls specify if you want multiple values on the value stack or packed into an array. The multiple-values-on-the-value-stack case matches the problem here.

**Comment by fatcerberus** [comment]
Hm, I'm not sure what a good neutral name would be that wouldn't be confusing, although I think I agree anyway that it makes the most sense to use the same constant for both cases.

Out of curiosity what happens right now if a Duktape/C function returns some positive value other than 1?

**Comment by svaarala** [comment]
An ApiError is thrown by call handling.
